### PR TITLE
Initial implementation for JSONFG output format

### PR DESCRIPTION
**Pull request by nls-jajuko**
hakunapi-jsonfg module 

First steps to support JSON-FG
https://docs.ogc.org/DRAFTS/21-045.html

### CODE REVIEW

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
+package fi.nls.hakunapi.geojson.hakuna;
+
+// TODO: forced to share package fi.nls.hakunapi.geojson.hakuna due to json.writeStringUnsafe scope
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java:3]: Let's just change that to `protected`, no real point in hiding it as the extending class has raw access to the actual `OutputStream` and can mess things up if it wants to.

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		final GeometryWriter geomWriter = propertyGeometryJson.computeIfAbsent(name, k -> new JSONFGGeometryWriter(json,
+				k.getBytes(StandardCharsets.UTF_8), forceLonLat || !crsIsLatLon, dims));
+		geometry.write(geomWriter);
+	}
+
+	@Override
+	public void startFeatureCollection(FeatureType ft, String layername) throws Exception {
+
+		collectionFt = ft;
+		if (isWGS84(outputSrid)) {
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java:75]: I guess we could just leave it up to ProjectionTransformerFactory to determine if toCRS84(srid) is NOP? That way JSONFG doesn't have to care about 84, 4326 or 4258 being WGS84.

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+		if (geometryDimension != null) {
+			json.writeFieldName(GEOMETRY_DIMENSION);
+			json.writeNumber(geometryDimension);
+		}
+		if (ftJSONFGSchema != null) {
+			json.writeStringField(FEATURE_SCHEMA, ftJSONFGSchema);
+		}
+
+		if (!isWgs84) {
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java:112]: If we let ProjectionTransformerFactory determine if the outputSrid is WGS84 compliant then we could use `outputWgs84Proj.isNOP()` to determine if `coordRefSys` needs to be written out.

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+			cachedGeometry = null;
+
+		} else {
+			// TODO check if this is required
+			json.writeFieldName(GEOMETRY);
+			json.writeNull();
+
+		}
+
+		// TODO what when why multiple temporal?
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java:214]: In `startFeatureCollection(FeatureType ft, String layername)` 
use `ft.getDatetimeProperties()` to get the datetime properties associated with time, cache their (public) names and on each `writeProperty(Instant | LocalDate)` check if the name is one of the configured ones and omit writing those out right away. Core 1 `?datetime=start/end` also gets compared against these properties.

Currently there's no way to tell if two configured time properties should be considered to form a time range, e.g.:
```
collections.my_collection.time=version_start_time,version_end_time
collections.my_collection.time.exclusive=version_end_time
```
but I think it's a safe assumption to make.
```
int size = ft.getDatetimeProperties().size();
if (size == 1) {
    // expect date/timestamp (check the property type for which to expect)
} else if (size == 2) {
  // expect and write out interval
}

**Comment by jampukka** [comment]
> "geometry" is reprojected within output format writer

I'm thinking we should probably add a default function to either `HakunaFeature` or `HakunaGeometry` which would provide the "lesser geometry" to be used as the value for `geometry` (like a simplified footprint or centroid of a building). The default implementation would return the default geometry as-is, but we could expand on that so that sources could provide simplified CRS84-ready versions of the geometry.

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java (hunk 1 above)
Comment (nls-jajuko) [reply src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java]: changed to protected 
moved jsonfg to fi.nls.hakunapi.jsonfg

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java (hunk 2 above)
Comment (nls-jajuko) [reply src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java]: retained srid==84 check for now NOP check did not work

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java (hunk 3 above)
Comment (nls-jajuko) [reply src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java]: retained srid==84 check for now NOP check did not work

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java (hunk 4 above)
Comment (nls-jajuko) [reply src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriter.java]: supports only one datetimeproperty at the moment

**Comment by nls-jajuko** [comment]
https://github.com/nlsfi/hakunapi/blob/main/src/hakunapi-simple-servlet-javax/src/main/java/fi/nls/hakunapi/simple/servlet/javax/GlobalFQueryParamFilter.java

I think GlobalFQueryParamFilter clashes with custom output formats

File: src/hakunapi-jsonfg/src/test/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriterTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+package fi.nls.hakunapi.geojson.hakuna;
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/test/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriterTest.java:1]: Package doesn't match that of `fi.nls.hakunapi.jsonfg.JSONFGFeatureCollectionWriter` which isn't conventional. Obviously package private/protected methods aren't visible to the unit test class but those don't seem to get tested.

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            geometry.write(geometryJson);
+        } else {
+            json.writeFieldName(GEOMETRY);
+            json.writeNull();
+        }
+    }
+
+    public static void writeTemporal(HakunaJsonWriter json, LocalDate date, Instant timestamp) throws IOException {
+        if (date == null && timestamp == null) {
+            return;
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java:104]: This is probably fine, hopefully JSONFG doesn't require `"time": null` if the value is null (like GeoJSON does with `"properties": null`)

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        json.writeEndObject();
+
+    }
+
+    public static HakunaGeometry getFootprintGeometry(HakunaGeometry placeGeometry,
+            ProjectionTransformer outputCrs84Proj) {
+
+        Geometry src = placeGeometry.toJTSGeometry();
+
+        Geometry dup = src.copy();
```
Comment (jampukka) [inline src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java:127]: Looking at this I should probably add heavy comments that informs others that `HakunaGeometry.toJTSGeometry()` and `ProjectionHelper.reproject()` are allowed to return `Geometry`s that might be backed by the same mutable data.

**Comment by jampukka** [comment]
Looks good to me

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        case GEOMETRYCOLLECTION:
+        default:
+            break;
+
+        }
+        return geometryDimension;
+    }
+
+    public static void writePlace(HakunaJsonWriter json, HakunaGeometry placeGeometry, HakunaGeometry geometry,
+            JSONFGGeometryWriter placeJson, GeometryWriter geometryJson, ProjectionTransformer outputCrs84Proj)
```
[the remark this answers is not in the document]
Comment (nls-jajuko) [reply src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java]: fixed - geometries produced out of writePlace

File: src/hakunapi-jsonfg/src/test/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriterTest.java (hunk 5 above)
Comment (nls-jajuko) [reply src/hakunapi-jsonfg/src/test/java/fi/nls/hakunapi/geojson/hakuna/JSONFGFeatureCollectionWriterTest.java]: fixed - forgot to rename test packages

File: src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java (hunk 6 above)
Comment (jampukka) [reply src/hakunapi-jsonfg/src/main/java/fi/nls/hakunapi/jsonfg/JSONFG.java:104]: Actually looking into this it looks like `"time"` property is required:
https://docs.ogc.org/DRAFTS/21-045.html#schema-feature
but value can be either null or object:
https://docs.ogc.org/DRAFTS/21-045.html#schema-time

So my guess is that this is against the schema and we should replace the early return with with writing `"time": null`

**Comment by nls-jajuko** [comment]
first steps towards validating JSONFG output - should find out a source for the jsonfg json-schemas

**Comment by nls-jajuko** [comment]
Added a generic validation test for feature collection writer

**Pull request merged**